File: app/src/main/java/com/engineeringforyou/basesite/StreetViewActivity.java
package com.engineeringforyou.basesite;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.ViewGroup.LayoutParams;

import com.google.android.gms.maps.StreetViewPanoramaOptions;
import com.google.android.gms.maps.StreetViewPanoramaView;
import com.google.android.gms.maps.model.LatLng;


public class StreetViewActivity extends AppCompatActivity {

//Define the LatLng value we’ll be using for the paranorma’s initial camera position//

    private static final LatLng LONDON = new LatLng(51.503324, -0.119543);
    private StreetViewPanoramaView mStreetViewPanoramaView;
    private static final String STREETVIEW_BUNDLE_KEY = "StreetViewBundleKey";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

//Configure the panorama by passing in a StreetViewPanoramaOptions object//

        StreetViewPanoramaOptions options = new StreetViewPanoramaOptions();
        if (savedInstanceState == null) {

//Set the panorama’s location//

            options.position(LONDON);
        }

        mStreetViewPanoramaView = new StreetViewPanoramaView(this, options);
        addContentView(mStreetViewPanoramaView,
                new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT));

        Bundle mStreetViewBundle = null;
        if (savedInstanceState != null) {
            mStreetViewBundle = savedInstanceState.getBundle(STREETVIEW_BUNDLE_KEY);
        }
        mStreetViewPanoramaView.onCreate(mStreetViewBundle);
    }

}

File: app/src/main/java/com/engineeringforyou/basesite/presentation/streetview/StreetViewActivity.kt
package com.engineeringforyou.basesite.presentation.streetview


import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import com.engineeringforyou.basesite.R
import com.engineeringforyou.basesite.models.Site
import com.engineeringforyou.basesite.utils.EventFactory
import com.google.android.gms.maps.StreetViewPanorama
import com.google.android.gms.maps.SupportStreetViewPanoramaFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.StreetViewPanoramaCamera
import com.google.android.gms.maps.model.StreetViewSource

class StreetViewActivity : AppCompatActivity() {

    companion object {
        private const val POINT = "point"

        fun start(activity: Activity, site: Site) {
            EventFactory.openStreetView(site)
            val intent = Intent(activity, StreetViewActivity::class.java)
            intent.putExtra(POINT, LatLng(site.latitude!!, site.longitude!!))
            activity.startActivity(intent)
            activity.overridePendingTransition(R.anim.slide_left_in, R.anim.alpha_out)
        }
    }

    private var point: LatLng? = null

    private lateinit var streetViewPanorama: StreetViewPanorama

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_street_view)

        val streetViewPanoramaFragment = supportFragmentManager
                .findFragmentById(R.id.streetviewpanorama) as SupportStreetViewPanoramaFragment
        streetViewPanoramaFragment.getStreetViewPanoramaAsync { panorama ->
            streetViewPanorama = panorama
            if (savedInstanceState == null) {
                point = intent.getParcelableExtra(POINT)
                streetViewPanorama.setPosition(point, StreetViewSource.OUTDOOR)
                setOrientation()
            }
        }
        initToolbar()
    }

    private fun initToolbar() {
        val actionBar = supportActionBar
        actionBar?.setDisplayShowHomeEnabled(false)
        actionBar?.setDisplayHomeAsUpEnabled(true)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun setOrientation() {
        if (streetViewPanorama.location == null) {
            Handler().postDelayed({ setOrientation() }, 750L)
            return
        }
        streetViewPanorama.animateTo(StreetViewPanoramaCamera.Builder().apply {
            zoom(zoom)
            tilt(30f)
            bearing(bearing(streetViewPanorama.location.position, point!!))
        }.build(), 300L)
    }

    private fun distance(center: LatLng, point: LatLng): Float {
        val R = 6371000
        val lat1 = Math.toRadians(center.latitude)
        val lon1 = Math.toRadians(center.longitude)
        val lat2 = Math.toRadians(point.latitude)
        val lon2 = Math.toRadians(point.longitude)

        val sin1 = Math.sin((lat1 - lat2) / 2)
        val sin2 = Math.sin((lon1 - lon2) / 2)
        return (2 * R * Math.asin(Math.sqrt(sin1 * sin1 + sin2 * sin2 * Math.cos(lat1) * Math.cos(lat2)))).toFloat()
    }

    private fun bearing(center: LatLng, point: LatLng): Float {
        val lon1 = center.longitude
        val lat1 = center.latitude
        val lon2 = point.longitude
        val lat2 = point.latitude

        val y = Math.sin(lon2 - lon1) * Math.cos(lat2)
        val x = Math.cos(lat1) * Math.sin(lat2) - Math.sin(lat1) * Math.cos(lat2) * Math.cos(lon2 - lon1)
        val bearing = Math.toDegrees(Math.atan2(y, x)).toFloat()
        return if (bearing < 0) 360 + bearing else bearing
    }

}